homeassistant/components/onvif/event.py

```diff
@@ -87,188 +124,676 @@ def async_remove_listener(self, update_callback: CALLBACK_TYPE) -> None:
+            self._event_manager.async_webhook_failed()
+            LOGGER.debug(
+                "%s: Device does not support notification service or too many subscriptions: %s",
+                self._name,
+                _stringify_onvif_error(err),
+            )
+            return False
 
-            topic = msg.Topic._value_1
-            if not (parser := PARSERS.get(topic)):
-                if topic not in UNHANDLED_TOPICS:
-                    LOGGER.info(
-                        "No registered handler for event from %s: %s",
-                        self.unique_id,
-                        msg,
-                    )
-                    UNHANDLED_TOPICS.add(topic)
-                continue
+        self._async_schedule_webhook_renew(SUBSCRIPTION_RENEW_INTERVAL)
+        return True
 
-            event = await parser(self.unique_id, msg)
+    async def _async_restart_webhook(self) -> bool:
+        """Restart the webhook subscription assuming the camera rebooted."""
+        await self._async_unsubscribe_webhook()
+        return await self._async_start_webhook()
 
-            if not event:
-                LOGGER.info("Unable to parse event from %s: %s", self.unique_id, msg)
+    async def _async_renew_webhook(self) -> bool:
+        """Renew webhook subscription."""
+        if not self._webhook_subscription:
+            return False
+        try:
+            await self._webhook_subscription.Renew(SUBSCRIPTION_RELATIVE_TIME)
+            LOGGER.debug("%s: Renewed Webhook subscription", self._name)
+            return True
+        except RENEW_ERRORS as err:
+            LOGGER.debug(
+                "%s: Failed to renew webhook subscription %s",
+                self._name,
+                _stringify_onvif_error(err),
+            )
+        return False
+
+    async def _async_renew_or_restart_webhook(
+        self, now: dt.datetime | None = None
+    ) -> None:
+        """Renew or start webhook subscription."""
+        if self._hass.is_stopping or self.state != WebHookManagerState.STARTED:
+            return
+        if self._renew_lock.locked():
+            LOGGER.debug("%s: Webhook renew already in progress", self._name)
+            # Renew is already running, another one will be
+            # scheduled when the current one is done if needed.
+            return
+        async with self._renew_lock:
+            next_attempt = SUBSCRIPTION_RENEW_INTERVAL_ON_ERROR
+            try:
+                if (
+                    await self._async_renew_webhook()
+                    or await self._async_restart_webhook()
+                ):
+                    next_attempt = SUBSCRIPTION_RENEW_INTERVAL
+            finally:
+                self._async_schedule_webhook_renew(next_attempt)
+
+    @callback
+    def _async_register_webhook(self) -> None:
+        """Register the webhook for motion events."""
+        LOGGER.debug("%s: Registering webhook: %s", self._name, self._webhook_unique_id)
+
+        try:
```

Comment:
Should this handle SSL as well? Not all camera's will be able to connect to SSL enabled endpoints. Home Assistant may be running using SSL (enabled in the http config).

Reply:
It will use pull point and only switch over to webhooks the first time that one is received so if ssl doesn’t work you get pull point instead

Reply:
https://github.com/home-assistant/core/pull/91485/files#diff-eb1134cbd2e70e5f98a0f81cfbec3adcfeae7e9c8c3817a640bc99d5742526eaR207

Reply:
Right, good (pull) point 👍